components/viaduct/src/backend/ffi.rs

```diff
@@ -37,160 +20,115 @@ macro_rules! backend_error {
-    // Note that these are guaranteed by
-    // https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html
-    // and thus this is a little paranoid.
-    ffi_support::static_assert!(
-        STATIC_ASSERT_USIZE_EQ_FUNC_SIZE,
-        std::mem::size_of::<usize>() == std::mem::size_of::<FetchCallback>()
-    );
-
-    ffi_support::static_assert!(
-        STATIC_ASSERT_USIZE_EQ_OPT_FUNC_SIZE,
-        std::mem::size_of::<usize>() == std::mem::size_of::<Option<FetchCallback>>()
-    );
+#[derive(Default)]
+pub struct FfiBackend;
 
-    /// Get the function pointer to the FetchCallback. Panics if the callback
-    /// has not yet been initialized.
-    pub(super) fn get_callback() -> Option<FetchCallback> {
-        let ptr_value = CALLBACK_PTR.load(Ordering::SeqCst);
-        unsafe { std::mem::transmute::<usize, Option<FetchCallback>>(ptr_value) }
+impl FfiBackend {
+    pub fn new() -> Self {
+        Self {}
     }
-
-    /// Set the function pointer to the FetchCallback. Returns false if we did nothing because the callback had already been initialized
-    pub(super) fn set_callback(h: FetchCallback) -> bool {
-        let as_usize = h as usize;
-        match CALLBACK_PTR.compare_exchange(0, as_usize, Ordering::SeqCst, Ordering::SeqCst) {
-            Ok(_) => true,
-            Err(_) => {
-                // This is an internal bug, the other side of the FFI should ensure
-                // it sets this only once. Note that this is actually going to be
-                // before logging is initialized in practice, so there's not a lot
-                // we can actually do here.
-                log::error!("Bug: Initialized CALLBACK_PTR multiple times");
-                false
-            }
+    pub fn set_backend(&self, fetch: Box<dyn Fetcher>) {
+        if FETCHER.set(fetch).is_err() {
+            log::warn!("viaduct http request backend already set");
         }
     }
 }
 
-/// Return a ByteBuffer of the requested size. This is used to store the
-/// response from the callback.
-#[no_mangle]
-pub extern "C" fn viaduct_alloc_bytebuffer(sz: i32) -> ByteBuffer {
-    let mut error = ffi_support::ExternError::default();
-    let buffer =
-        ffi_support::call_with_output(&mut error, || ByteBuffer::new_with_size(sz.max(0) as usize));
-    error.consume_and_log_if_error();
-    buffer
-}
-
-#[no_mangle]
-pub extern "C" fn viaduct_log_error(s: FfiStr<'_>) {
-    let mut error = ffi_support::ExternError::default();
-    ffi_support::call_with_output(&mut error, || {
-        log::error!("Viaduct Ffi Error: {}", s.as_str())
-    });
-    error.consume_and_log_if_error();
-}
-
-#[no_mangle]
-pub extern "C" fn viaduct_initialize(callback: FetchCallback) -> u8 {
-    ffi_support::abort_on_panic::call_with_output(|| callback_holder::set_callback(callback))
+impl Backend for FfiBackend {
+    fn send(&self, request: crate::Request) -> Result<crate::Response, Error> {
+        super::note_backend("FFI (trusted)");
```

Comment:
I'm starting to see where the `Ffi` prefix came from, but I'm a little confused. We have a backend called "FFI", and the structs we have in the udl have `Ffi` as a prefix, but *all* backends return instances of these Ffi structs?

I wonder if the default backend should be called either `default` or `internal`? I could well be missing something here though

Reply:
Yeah... I don't like the `Ffi` prefix on the UDL actually - I'll change that up!